### PR TITLE
8256858: C2: Devirtualize PhaseIterGVN-specific methods

### PR DESCRIPTION
**Pull request by cl4es**
PhaseValues define the virtual method is_IterGVN, which is trivially returning 0(!) for all types except those derived from PhaseIterGVN. Similarly there's igvn_rehash_node_delayed which is virtual and a no-op in base types, but implemented to call rehash_node_delayed in PhaseIterGVN.

By devirtualizing we allow for more aggressive inlining and slightly better code generation in a few places. 

This increases sizeof(PhaseValues) from 2480 to 2488 on x64. Since we only go through a limited number of phases per compilation this seems acceptable.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256858](https://bugs.openjdk.java.net/browse/JDK-8256858): C2: Devirtualize PhaseIterGVN-specific methods

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to ffae8ad9eebfab0a936911c24f3d7e3bfc44dddb
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1385/head:pull/1385`
`$ git checkout pull/1385`

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/opto/addnode.cpp [hunk 1]
```diff
@@ -191,10 +191,9 @@ Node *AddNode::Ideal(PhaseGVN *phase, bool can_reshape) {
       set_req(1, addx);
       set_req(2, a22);
       progress = this;
-      PhaseIterGVN *igvn = phase->is_IterGVN();
-      if (add2->outcnt() == 0 && igvn) {
+      if (add2->outcnt() == 0 && phase->is_IterGVN()) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/addnode.cpp:194]: Why did you remove the local variable `igvn`? Performance wise it shouldn't make a difference, right?

File: src/hotspot/share/opto/addnode.cpp [hunk 2]
```diff
@@ -627,10 +626,9 @@ Node *AddPNode::Ideal(PhaseGVN *phase, bool can_reshape) {
     if( t22->singleton() && (t22 != Type::TOP) ) {  // Right input is an add of a constant?
       set_req(Address, phase->transform(new AddPNode(in(Base),in(Address),add->in(1))));
       set_req(Offset, add->in(2));
-      PhaseIterGVN *igvn = phase->is_IterGVN();
-      if (add->outcnt() == 0 && igvn) {
+      if (add->outcnt() == 0 && phase->is_IterGVN()) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/addnode.cpp:629]: Again, I think replacing `igvn` by `phase->is_IterGVN()` makes the code less readable.

File: src/hotspot/share/opto/memnode.cpp [hunk 3]
```diff
@@ -2746,14 +2748,14 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
     }
   }
 
-  if (result != this && phase->is_IterGVN() != NULL) {
+  PhaseIterGVN* igvn = phase->is_IterGVN();
+  if (result != this && igvn != NULL) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/memnode.cpp:2752]: Here you did the reverse (replaced `phase->is_IterGVN()` by `igvn` local).

File: src/hotspot/share/opto/phaseX.hpp [hunk 4]
```diff
@@ -369,12 +366,12 @@ class PhaseTransform : public Phase {
...
+  bool      _iterGVN;
 public:
-  PhaseValues( Arena *arena, uint est_max_size );
-  PhaseValues( PhaseValues *pt );
-  NOT_PRODUCT( ~PhaseValues(); )
-  virtual PhaseIterGVN *is_IterGVN() { return 0; }
+  PhaseValues(Arena* arena, uint est_max_size);
+  PhaseValues(PhaseValues* pt);
+  NOT_PRODUCT(~PhaseValues();)
+  PhaseIterGVN *is_IterGVN() { return (_iterGVN) ? (PhaseIterGVN*)this : NULL; }
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/phaseX.hpp:374]: `PhaseIterGVN *is_IterGVN` -> `PhaseIterGVN* is_IterGVN`

File: src/hotspot/share/opto/phaseX.hpp [hunk 5]
```diff
@@ -391,7 +388,7 @@ class PhaseValues : public PhaseTransform {
 
   virtual ConNode* uncached_makecon(const Type* t);  // override from PhaseTransform
 
-  virtual const Type* saturate(const Type* new_type, const Type* old_type,
+  const Type* saturate(const Type* new_type, const Type* old_type,
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/phaseX.hpp:391]: Indentation should be fixed.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

**Review by TobiHartmann — APPROVED** [review]

**Comment by cl4es** [comment]
@vnkozlov @TobiHartmann - thank you for reviewing!

/integrate

**Pull request closed without merging**